feat: improve HTTP redirects behavior

pyrooka opened the pull request:
This commit modifies the Node core so that it will include "safe" headers when performing a cross-site redirect where both the original and redirected hosts are within IBM's "cloud.ibm.com" domain.

<!--
Thank you for your pull request!

Please provide a description above and review the requirements below.

Bug fixes and new features should include tests whenever possible.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `npm test` passes (tip: `npm run lint-fix` can correct most style issues)
- [x] tests are included
- [x] documentation is changed or added

pyrooka committed:
feat: improve HTTP redirects behavior

This commit modifies the Node core so that it will include "safe" headers
when performing a cross-site redirect where both the original and redirected
hosts are within IBM's "cloud.ibm.com" domain.

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

pyrooka committed:
chore: update .secrets.baseline

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

pyrooka committed:
chore: fix command in the GitHub PR template

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

padamstx commented on .github/pull_request_template.md:
```diff
@@ -9,6 +9,6 @@ Bug fixes and new features should include tests whenever possible.
 <!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
 
-- [ ] `npm test` passes (tip: `npm run lint-fix` can correct most style issues)
+- [ ] `npm test` passes (tip: `npm run lint:fix` can correct most style issues)
```
good catch!

padamstx commented on lib/request-wrapper.ts:
```diff
@@ -580,3 +609,8 @@ function ensureJSONResponseBodyIsObject(response: any): any | string {
+
+// Returns true iff safe headers should be copied to a redirected request.
+function shouldCopySafeHeadersOnRedirect(fromHost: string, toHost: string): boolean {
+  return fromHost.endsWith('.cloud.ibm.com') && toHost.endsWith('.cloud.ibm.com');
```
This logic should work as long as we are also supporting the scenario where a redirection to the same host also results in the safe headers being copied, which I assume is the case since you're not completely overriding axios' redirection behavior (presumably) :)

pyrooka replied:
Those scenarios are covered by the unit tests now and yes, we are not overriding the whole redirection behavior, it's just a sort of post-processing in the chain, before sending the next, redirected request.

padamstx commented on test/unit/redirect.test.js:
```diff
@@ -0,0 +1,211 @@
+  });
+
+  it('should include safe headers within cloud.ibm.com domain', async () => {
+    const url1 = 'http://region1.cloud.ibm.com';
```
let's use https in these URLs :)

pyrooka replied:
Agreed! I also change the protocols in my Python core PR, because I used HTTP there too.

padamstx commented on test/unit/redirect.test.js:
```diff
@@ -0,0 +1,211 @@
+/**
```
need to add tests for redirects to the same host:
* both are region1.cloud.ibm.com
* both are region2.notcloud.ibm.com

padamstx reviewed (commented):
Looks good.
Only suggestion is to add a couple tests

pyrooka committed:
chore: test improvements

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

pyrooka committed:
chore: revamp unit tests

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

chore: small improvements

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

chore: fix tests

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

chore: fix unit tests

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

chore: lint fix

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

pyrooka committed:
chore: update `nock` to latest

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

pyrooka committed:
fix: use the correct conditions when copying safe headers

Signed-off-by: Norbert Biczo <pyrooka@users.noreply.github.com>

pyrooka committed:
Merge branch 'main' into nb/redirect-improvements
